src/platform/backends/qemu/dnsmasq_server.cpp

```diff
@@ -28,39 +31,24 @@ namespace mpl = multipass::logging;
 
 namespace
 {
-auto start_dnsmasq_process(const QDir& data_dir, const QString& bridge_name, const mp::IPAddress& bridge_addr,
-                           const mp::IPAddress& start_ip, const mp::IPAddress& end_ip)
+auto make_dnsmasq_process(const mp::ProcessFactory* process_factory, QDir data_dir, const QString& bridge_name,
+                          const mp::IPAddress& bridge_addr, const mp::IPAddress& start, const mp::IPAddress& end)
 {
-    auto cmd = std::make_unique<QProcess>();
-
-    QObject::connect(cmd.get(), &QProcess::readyReadStandardError,
-                     [&cmd]() { mpl::log(mpl::Level::error, "dnsmasq", cmd->readAllStandardError().data()); });
-
-    cmd->start("dnsmasq",
-               QStringList() << "--keep-in-foreground"
-                             << "--strict-order"
-                             << "--bind-interfaces"
-                             << "--except-interface=lo" << QString("--interface=%1").arg(bridge_name)
-                             << QString("--listen-address=%1").arg(QString::fromStdString(bridge_addr.as_string()))
-                             << "--dhcp-no-override"
-                             << "--dhcp-authoritative"
-                             << QString("--dhcp-leasefile=%1").arg(data_dir.filePath("dnsmasq.leases"))
-                             << QString("--dhcp-hostsfile=%1").arg(data_dir.filePath("dnsmasq.hosts")) << "--dhcp-range"
-                             << QString("%1,%2,infinite")
-                                    .arg(QString::fromStdString(start_ip.as_string()))
-                                    .arg(QString::fromStdString(end_ip.as_string())));
-
-    cmd->waitForStarted();
-    return cmd;
+    auto process_spec = std::make_unique<mp::DNSMasqProcessSpec>(data_dir, bridge_name, bridge_addr, start, end);
+    return process_factory->create_process(std::move(process_spec));
 }
 } // namespace
 
-mp::DNSMasqServer::DNSMasqServer(const Path& path, const QString& bridge_name, const IPAddress& bridge_addr,
-                                 const IPAddress& start, const IPAddress& end)
+mp::DNSMasqServer::DNSMasqServer(const ProcessFactory* process_factory, const Path& path, const QString& bridge_name,
+                                 const IPAddress& bridge_addr, const IPAddress& start, const IPAddress& end)
     : data_dir{QDir(path)},
-      dnsmasq_cmd{start_dnsmasq_process(data_dir, bridge_name, bridge_addr, start, end)},
+      dnsmasq_cmd{make_dnsmasq_process(process_factory, data_dir, bridge_name, bridge_addr, start, end)},
       bridge_name{bridge_name}
 {
+    QObject::connect(dnsmasq_cmd.get(), &Process::readyReadStandardError,
```

Comment:
Would it make sense to integrate that into Process? It feels like we often would want to log stderr of launched processes?

Reply:
It would, the only niggle is that desired the log level differs depending on the program (i.e. currently dnsmasq's stderr is logged as Error, but qemu's is logged as warning).

This also raises the question, should stdout be logged then too? Perhaps yes, with a non-default-displayed log level?

Reply:
I've added a error_log_level specifier to ProcessSpec, so each process can set how critical the child process' stderr should appear in the log

Reply:
> This also raises the question, should stdout be logged then too? Perhaps yes, with a non-default-displayed log level?

Yeah, all stdout should be logged to `debug` level I'd say and all stderr to `warning` by default.

Reply:
The `stdout` handling can come later though.